lib/ansible/modules/network/frr/frr_facts.py

```diff
@@ -0,0 +1,470 @@
+            return match.group(1)
+
+    def _parse_holding_block_headers(self, data):
+        match = re.search(r'Holding block headers:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_used_small_blocks(self, data):
+        match = re.search(r'Used small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_used_ordinary_blocks(self, data):
+        match = re.search(r'Used ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_free_small_blocks(self, data):
+        match = re.search(r'Free small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_free_ordinary_blocks(self, data):
+        match = re.search(r'Free ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_ordinary_blocks(self, data):
+        match = re.search(r'Ordinary blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_small_blocks(self, data):
+        match = re.search(r'Small blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_holding_blocks(self, data):
+        match = re.search(r'Holding blocks:(?:\s*)(.*)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def _parse_daemons(self, data):
+        match = re.search(r'Memory statistics for (\w+)', data, re.M)
+        if match:
+            return match.group(1)
+
+    def gather_memory_facts(self, data):
+        mem_details = data.split('\n\n')
+        mem_stats = {}
+        mem_counters = ['total_heap_allocated', 'holding_block_headers', 'used_small_blocks',
+                        'used_ordinary_blocks', 'free_small_blocks', 'free_ordinary_blocks',
+                        'ordinary_blocks', 'small_blocks', 'holding_blocks']
+
+        for item in mem_details:
+            daemon = self._parse_daemons(item)
+            mem_stats[daemon] = {}
+            for x in mem_counters:
+                meth = getattr(self, '_parse_%s' % x, None)
+                mem_stats[daemon][x] = meth(item)
+
+        return mem_stats
+
+    def populate(self):
+        super(Hardware, self).populate()
+        data = self.responses[0]
+        if data:
+            self.facts['mem_stats'] = self.gather_memory_facts(data)
+
+
+class Config(FactsBase):
+
+    COMMANDS = ['show running-config']
+
+    def populate(self):
+        super(Config, self).populate()
+        data = self.responses[0]
+        if data:
+            data = re.sub(r'^Building configuration...\s+Current configuration:', '', data, flags=re.MULTILINE)
+            self.facts['config'] = data
+
+
+class Interfaces(FactsBase):
+
+    COMMANDS = ['show interface']
+
+    def populate(self):
+        ldp_supported = get_capabilities(self.module)['supported_protocols']['ldp']
```

Comment:
Instead of calling `get_capabilities` twice maybe make it an attribute for `FactsBase` class in the `__init__` method? 
`self._capabilities = get_capabilities(self._module)`

Reply:
Agreed.